File: DNA.java
// time complexity :O(n)
// space complextiy : O(n)

class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        HashMap<String, Integer> map = new HashMap<>();
        List<String> result = new ArrayList<>();
        int n = s.length();
        // base case 
        if(n < 10) 
            return result;
        // else
        for(int i = 0; i<n-9; i++)
            map.put(s.substring(i,i+10),map.getOrDefault(s.substring(i,i+10), 0)+1);
        for(String word : map.keySet())
            if(map.get(word)>1)
                result.add(word);
        return result;
    }
}

File: DNA.cpp
/*
We use a sliding window (2 pointer technique) to get all possible substrings.
We store it in a set. If they repeat again we put it in a result.
Time Complexity: O(N*L), Here L = 10.
Space Compleixty: O(N*L), All strings are unique.
*/

class Solution {
public:
    vector<string> findRepeatedDnaSequences(string s) {
        set<string>resultSet;
        set<string>potentialString;
        string newString = "";
        for ( int i = 0; i+10 <= s.size(); i++){
            newString = s.substr(i,10);
            cout<<newString<<endl;
            if ( potentialString.find(newString) != potentialString.end()){
                resultSet.insert(newString);
            }
            potentialString.insert(newString);
            
        }
        vector<string>result;
        for ( auto val: resultSet){
            result.push_back(val);
            
        }
        return result;
    }
};